app/extensions/tus/__init__.py

```diff
@@ -68,29 +80,218 @@ def _tus_file_handler(upload_file_path, filename, original_filename, req, app):
+        with open(metadata_filepath, 'w') as metadata_file:
+            metadata = {
+                'user_guid': None
+                if current_user.is_anonymous
+                else str(current_user.guid),
+            }
+            json.dump(metadata, metadata_file)
+
     log.debug('Tus finished uploading: %r in dir %r.' % (filename, dir))
     filepath = os.path.join(dir, filename)
-    os.rename(upload_file_path, filepath)
 
-    # Store the original filename as metadata next to the file
-    tus_write_file_metadata(filepath, original_filename)
+    try:
+        os.rename(upload_file_path, filepath)
+
+        # Store the original filename as metadata next to the file
+        tus_write_file_metadata(filepath, original_filename, resource_id)
+    except Exception:
+        if os.path.exists(filepath):
+            os.rename(filepath, upload_file_path)
+        raise
 
     return filename
 
 
-def tus_get_metadata_filepath(filepath):
+def _tus_delete_file_handler(upload_file_path, resource_id, req, app):
+    from uuid import UUID
+
+    asset_group_id = None
+    transaction_id = None
+    if 'x-houston-asset-group-id' in req.headers:
+        asset_group_id = req.headers.get('x-houston-asset-group-id')
+        UUID(
+            asset_group_id, version=4
+        )  # this is just to test it is a valid uuid - will throw ValueError if not! (500 response)
+    # TODO verify asset_group? ownership? etc  possibly also "session-secret" to prevent anyone from adding to asset_group
+    if 'x-tus-transaction-id' in req.headers:
+        transaction_id = req.headers.get('x-tus-transaction-id')
+        UUID(
+            transaction_id, version=4
+        )  # this is just to test it is a valid uuid - will throw ValueError if not! (500 response)
+
+    dir = os.path.join(tus_upload_dir(app), 'unknown')
+    if asset_group_id is not None:
+        dir = tus_upload_dir(app, git_store_guid=asset_group_id)
+    elif transaction_id is not None:
+        dir = tus_upload_dir(app, transaction_id=transaction_id)
+    elif 'session' in req.cookies:
+        dir = tus_upload_dir(
+            app, session_id=str(req.cookies.get('session')).encode('utf-8')
+        )
+
+    filepaths = []
+    # traverse whole upload dir and take everything
+    for root, dirs, files in os.walk(dir):
+        for path in files:
+            if not path.startswith('.'):
+                filepaths.append(os.path.join(root, path))
+
+    matches = []
+    for filepath in filepaths:
+        assert os.path.exists(filepath)
+        metadata_filepath = tus_get_resource_metadata_filepath(filepath)
+        if os.path.exists(metadata_filepath):
+            with open(metadata_filepath, 'r') as metadata_file:
+                metadata = json.load(metadata_file)
+                if resource_id == metadata.get('resource_id'):
+                    matches.append(filepath)
+                    matches.append(metadata_filepath)
+
+    log.debug(
```

Comment:
Deleting data should probably be recorded as an audit log too.

Reply:
Switched to using AuditLog